src/lib.rs

```diff
@@ -3133,6 +3135,49 @@ pub trait Itertools: Iterator {
         self.k_largest_by(k, k_smallest::key_to_cmp(key))
     }
 
+    /// Consumes the iterator and return an iterator of the last `n` elements.
+    ///
+    /// It allocates up to `n` elements.
+    /// The iterator, if directly collected to a `Vec`, is converted
+    /// without any extra copying or allocation cost.
+    ///
+    /// ```
+    /// use itertools::{assert_equal, Itertools};
+    ///
+    /// let v = vec![5, 9, 8, 4, 2, 12, 0];
+    /// assert_equal(v.iter().tail(3), &[2, 12, 0]);
+    /// assert_equal(v.iter().tail(10), &v);
+    ///
+    /// assert_equal((0..100).tail(10), 90..100);
+    /// ```
+    ///
+    /// For double ended iterators without side-effects, you might prefer
+    /// `.rev().take(n).rev()` to have a similar result (lazy and non-allocating)
+    /// without consuming the entire iterator.
+    #[cfg(feature = "use_alloc")]
+    fn tail(self, n: usize) -> VecDequeIntoIter<Self::Item>
+    where
+        Self: Sized,
+    {
+        match n {
+            0 => {
+                self.last();
+                VecDeque::new()
+            }
+            1 => self.last().into_iter().collect(),
+            _ => {
+                let mut iter = self.fuse();
+                let mut data: VecDeque<_> = iter.by_ref().take(n).collect();
+                iter.for_each(|value| {
+                    data.pop_front();
+                    data.push_back(value);
```

Comment:
I hoped that there would be a method that avoids the intermediate bookkeeping of pop_front/push_back (namely, updating `len`), but I did not find one, so that's fine with me.

However, @scottmcm's go-ahead would be appreciated here.

Reply:
I executed the benchmark I wrote earlier (`v.iter().tail(10)` with `let v = (0..1024).collect_vec();`) and the `VecDeque` version is 4 times slower.
Not sure why without investigation but apart from updating `len`, it also checks `is_empty` and `is_full` while the `Vec` version may just be more straightforward.

Reply:
And `data.rotate_left(1); *data.back_mut().unwrap() = value;` is no better.

Reply:
> I executed the benchmark I wrote earlier (`v.iter().tail(10)` with `let v = (0..1024).collect_vec();`) and the `VecDeque` version is 4 times slower. Not sure why without investigation but apart from updating `len`, it also checks `is_empty` and `is_full` while the `Vec` version may just be more straightforward.

Good catch. I guess it depends on the iterator length and the number of elements you collect. (When collecting the 10-tail of 1024 elements, maybe the iterations dominate the runtime, so that a `Vec` is better there and the final `rotate_left` does not matter that much?)

Under your findings, and unless we have evidence that `VecDeque` is better in realistic circumstances, I'd go with the simpler `Vec`, even accepting the breaking change if we one day discover the contrary.

Reply:
I think people would mostly want a small tail. If they want most of the iterator, `collect_vec+truncate` is simpler.

After running cargo-flamegraph on the benchmark, I don't really know more: it spends most of the time in `push_back` and a bit (7 times less) in `pop_front`, nothing else matters. So the iteration dominate runtime.

Maybe I would have better understanding with _godbolt_. I think my straightforward `Vec` version is better optimized than the `VecDeque` one.

---

While we don't do const-generics yet, the `Vec` version would be easily generalizable to both cases because update buffers (`[_; N]` and `Vec<_>`) is update a slice while the `VecDeque` version hide details.

Reply:
Your "skip" idea is _awesome_ in the case the size hint knows enough but it does not change a thing otherwise.
This second idea of yours reduce the impact of the final `rotate_left`, nice!

And if there was a `pop_front_push_back` in the future (as efficient as my cyclic update of the vector), we could revisit this, collect to VecDeque, then `iter.for_each(|val| { data.pop_front_push_back(val); })`.